---
title: "Introduction to the Tidyverse"
subtitle: "wrangling data with dplyr"
author: "Malcolm Barrett"
date: "June 18, 2019"
output:
  xaringan::moon_reader:
    css: ["default", "theme.css"]
    lib_dir: libs
    nature:
      highlightStyle: github
      highlightLines: true
      countIncrementalSlides: false
---

```{r setup, include=FALSE}
options(htmltools.dir.version = FALSE, tibble.max_extra_cols = 6, tibble.width = 60)
knitr::opts_chunk$set(warning = FALSE, message = FALSE, fig.align = "center", dpi = 320)
library(mbmisc)
library(tidyverse)
library(tidymeta)
```

class: inverse, center

background-image: url(../Images/tidyhex.png)
---

class: inverse, center

# Working with data in R
## the `r font_col("tidyverse", "#E69F00")` is a collection of `r font_col("friendly and consistent", "#56B4E9")` tools for data analysis and visualization. 

---
class: inverse, center

# Working with data in R
## `r font_col("the tidyverse is a collection of friendly and consistent tools for data analysis and visualization. ", "#6C7B7F")`
## They live as `r font_col("R packages", "#E69F00")` each of which does one thing well. 

---

background-image: url(https://raw.githubusercontent.com/tidyverse/tidyverse/master/man/figures/logo.png)
background-position: 90% 10%
## `library(tidyverse)` will load  
##  core tidyverse packages: 

#### [ggplot2](http://ggplot2.tidyverse.org), for data visualisation.
#### [dplyr](http://dplyr.tidyverse.org), for data manipulation.
#### [tidyr](http://tidyr.tidyverse.org), for data tidying.
#### [readr](http://readr.tidyverse.org), for data import.
#### [purrr](http://purrr.tidyverse.org), for functional programming.
#### [tibble](http://tibble.tidyverse.org), for tibbles, a modern re-imagining of data frames.
#### [stringr](https://github.com/tidyverse/stringr), for strings.
#### [forcats](https://github.com/hadley/forcats), for factors.

---

class: inverse, center

# Game plan:
## 8:30-8:35: importing data: `readr` and `haven`
## 8:35-9:30: transforming data: `dplyr` 
## 9:30-10:30: visualizing data: `ggplot2`
## 10:30-10:45: break
## 10:45-11:45: data viz exercise: `ggplot2`, `dplyr`, and friends
## 11:45-12:30: reproducible reports: `R Markdown`

---

class: inverse, center

# Game plan:
## **`readr/haven`**
## transforming data: `dplyr`
## visualizing data: `ggplot2`
## data viz exercise
## reproducible reports: `R Markdown`

---

background-image: url(http://hexb.in/hexagons/readr.png)
background-position: 93% 10%

## `r font_col("readr", "#E69F00")`

Function       | Reads
-------------- | --------------------------
`read_csv()`   | Comma separated values
`read_csv2()`  | Semi-colon separate values
`read_delim()` | General delimited files
`read_fwf()`   | Fixed width files
`read_log()`   | Apache log files
`read_table()` | Space separated files
`read_tsv()`   | Tab delimited values

---
## Importing Data

```{r eval = FALSE}
dataset <- read_csv("file_name.csv")
dataset
```

---

```{r}
diabetes <- read_csv("../Data/diabetes.csv")
diabetes
```
---

## Tibbles

## `r font_col("<code>data.frames</code>", "#E69F00")` are the basic form of rectangular data in R (columns of variables, rows of observations)

---

## Tibbles

## `r font_col("<code>data.frames</code> are the basic form of rectangular data in R (columns of variables, rows of observations", "#E5E5E5")`

## `read_csv()` reads the data into a `r font_col("tibble", "#E69F00")`, a modern version of the data frame. 

---

## Tibbles

## `r font_col("<code>data.frames</code> are the basic form of rectangular data in R (columns of variables, rows of observations", "#E5E5E5")`

## `r font_col("<code>read_csv()</code> reads the data into a tibble, a modern version of the data frame.", "#E5E5E5")`

## a tibble `r font_col("is", "#E69F00")` a data frame

---

background-image: url(http://hexb.in/hexagons/haven.png)
background-position: 90% 10%

## `r font_col("haven", "#E69F00")`

Function       | Software
-------------- | --------------------------
`read_sas()`   | SAS
`read_xpt()`   | SAS
`read_spss()`  | SPSS
`read_sav()`   | SPSS
`read_por()`   | SPSS
`read_stata()` | Stata
`read_dta()`   | Stata

## haven is *not* a core member of the tidyverse. That means you need to load it with `library(haven)`.

---

```{r}
library(haven)
diabetes <- read_sas("../Data/diabetes.sas7bdat")
diabetes
```

---

## Writing data

Function            | Writes
------------------- | ----------------------------------------
`write_csv()`       | Comma separated values
`write_excel_csv()` | CSV that you plan to open in Excel
`write_delim()`     | General delimited files
`write_file()`      | A single string, written as is
`write_lines()`     | A vector of strings, one string per line
`write_tsv()`       | Tab delimited values
`write_rds()`       | A data type used by R to save objects
`write_sas()`       | SAS .sas7bdat files
`write_xpt()`       | SAS transport format, .xpt
`write_sav()`       | SPSS .sav files
`write_stata()`     | Stata .dta files

```{r eval = FALSE}
write_csv(diabetes, path = "diabetes-clean.csv")
```

---

class: inverse, center

# Game plan:
## importing data: `readr`/`haven`
## **transforming data: `dplyr`**
## visualizing data: `ggplot2`
## data viz exercise
## reproducible reports: `R Markdown`

---
background-image: url(http://hexb.in/hexagons/dplyr.png)
background-position: 90% 10%

```{r transform, include=FALSE}
library(tidyverse)
library(gapminder)
```

## The main verbs of `dplyr`

## `select()`

## `filter()`

## `mutate()`

## `arrange()`

## `summarize()`

## `group_by()`

---

## The main verbs of `dplyr`

## `r font_col("<code>select()</code>", "#E69F00")` = `r font_col("Subset columns (variables)", "#56B4E9")`

## `filter()`

## `mutate()`

## `arrange()`

## `summarize()`

## `group_by()`


---

## `select()`

```{r, eval=FALSE}
select(<DATA>, <VARIABLES>)
```

---

## `select()`

```{r, eval=FALSE}
select(<DATA>, <VARIABLES>)
```

```{r, highlight.output = c(2, 14)} 
diamonds
```

---

## `select()`

```{r, eval = FALSE} 
select(diamonds, carat, cut, color, clarity) #<<
```

---

## `select()`

```{r, highlight.output = 2} 
select(diamonds, carat, cut, color, clarity)
```

---

## `select()`

```{r, eval=FALSE} 
select(diamonds, carat, cut, color, clarity)

select(diamonds, carat:clarity)

select(diamonds, 1:4)

select(diamonds, starts_with("c"))

?select_helpers
```

---

## gapminder

```{r}
library(gapminder)
gapminder
```

---

## Your turn 1

## Alter the code to select just the `pop` column:

```{r, eval=FALSE}
select(gapminder, year, lifeExp)
```

---

## Your Turn 1

```{r}
select(gapminder, pop)
```

---

## Quiz

Which of these is NOT a way to select the `country` and `continent` columns together?

```{r, eval=FALSE}
select(gapminder, -c(year, lifeExp, pop, gdpPercap))

select(gapminder, country:continent)

select(gapminder, starts_with("c"))

select(gapminder, ends_with("t"))
```

---

## Quiz

Which of these is NOT a way to select the `country` and `continent` columns together?

```{r, highlight.output = 2}
select(gapminder, ends_with("t"))
```

---

## The main verbs of `dplyr`

## `select()`

## `r font_col("<code>filter()</code>", "#E69F00")` = `r font_col("Subset rows by value", "#56B4E9")`

## `mutate()`

## `arrange()`

## `summarize()`

## `group_by()`

---

## `filter()`

```{r, eval=FALSE}
filter(<DATA>, <PREDICATES>)
```

--

### *Predicates: `TRUE/FALSE` statements*

---

## `filter()`

```{r, eval=FALSE}
filter(<DATA>, <PREDICATES>)
```

### ~~Predicates: `TRUE/FALSE` statements~~

### *Comparisons: `>`, `>=`, `<`, `<=`, `!=` (not equal), and `==` (equal).*

---

## `filter()`

```{r, eval=FALSE}
filter(<DATA>, <PREDICATES>)
```

### ~~Predicates: `TRUE/FALSE` statements~~

### ~~Comparisons: `>`, `>=`, `<`, `<=`, `!=` (not equal), and `==` (equal).~~

### *Operators: `&` is "and", `|` is "or", and `!` is "not"*

---

## `filter()`

```{r, eval=FALSE}
filter(<DATA>, <PREDICATES>)
```

### ~~Predicates: `TRUE/FALSE` statements~~

### ~~Comparisons: `>`, `>=`, `<`, `<=`, `!=` (not equal), and `==` (equal).~~

### ~~Operators: `&` is "and", `|` is "or", and `!` is "not"~~

### *`%in%`*

--

```{r} 
"a" %in% c("a", "b", "c")
```

---

## `filter()`

```{r, eval = FALSE} 
filter(diamonds, cut == "Ideal", carat > 3)
```

---

## `filter()`

```{r, highlight.output = c(1, 4:7)} 
filter(diamonds, cut == "Ideal", carat > 3)
```

---

## Your turn 2

### Show:

### All of the rows where `pop` is greater than or equal to 100000  
### All of the rows for El Salvador  
### All of the rows that have a missing value for `year`  (no need to edit this code)

---

## Your turn 2

### Show:

### All of the rows where `pop` is greater than or equal to 100000  
### All of the rows for El Salvador  
### All of the rows that have a missing value for `year`  (no need to edit this code)

```{r, eval=FALSE}
filter(gapminder, pop >= 100000)
filter(gapminder, country == "El Salvador")
filter(gapminder, is.na(year))
```

---

## `filter()`

```{r}
filter(diamonds, cut == "Ideal" | cut == "Very Good", carat > 3)
```

---

## Your turn 3

### Use Boolean operators to alter the code below to return only the rows that contain:

### El Salvador  
### Countries that had populations over 100000 in 1960 or earlier

```{r, eval=FALSE}
filter(gapminder, country == "El Salvador" | country == "Oman")
filter(______, _______)
```

---

## Your turn 3

### Use Boolean operators to alter the code below to return only the rows that contain:

### El Salvador  
### Countries that had populations over 100000 in 1960 or earlier

```{r, eval=FALSE}
filter(gapminder, country == "El Salvador")
filter(gapminder, pop > 100000, year <= 1960)
```

---

## The main verbs of `dplyr`

## `select()`

## `filter()`

## `r font_col("<code>mutate()</code>", "#E69F00")` = `r font_col("Change or add a variable", "#56B4E9")`

## `arrange()`

## `summarize()`

## `group_by()`
---

## `mutate()`

```{r, eval=FALSE}
mutate(<DATA>, <NAME> = <FUNCTION>)
```

---

## `mutate()`

```{r, eval = FALSE}
mutate(diamonds, log_price = log(price), log_pricesq = log_price^2)
```

---

## `mutate()`

```{r, highlight.output = 15}
mutate(diamonds, log_price = log(price), log_pricesq = log_price^2)
```

---

## The main verbs of `dplyr`

## `select()`

## `filter()`

## `mutate()`

## `r font_col("<code>arrange()</code>", "#E69F00")` = `r font_col("Sort the data set", "#56B4E9")`

## `summarize()`

## `group_by()`
---

## `arrange()`

```{r, eval=FALSE}
arrange(<DATA>, <SORTING VARIABLE>)
```

---

## `arrange()`

```{r}
arrange(diamonds, price)
```

---

## `arrange()`

```{r}
arrange(diamonds, cut, price)
```

---

## `desc()`

```{r}
arrange(diamonds, cut, desc(price))
```

---

## Your turn 4

## Arrange gapminder by `year`. Add `lifeExp` as a second (tie breaking) variable to arrange on. 

## Which country had the lowest life expectancy in 1952?

---

## Your turn 4

```{r}
arrange(gapminder, year, lifeExp)
```

---

## Your turn 5

## Use `desc()` to find the country with the highest `gdpPercap`.

---

## Your turn 5

```{r}
arrange(gapminder, desc(gdpPercap))
```

---
class: inverse, center

# Detour: The Pipe

# `r font_col("<code>%>%</code>", "#E69F00")`
## Passes the result on one function to another function

---

## Detour: The Pipe

```{r, eval=FALSE}
diamonds <- arrange(diamonds, price)
diamonds <- filter(diamonds, price > 300)
diamonds <- mutate(diamonds, log_price = log(price))

diamonds
```

---

## Detour: The Pipe

```{r, eval=FALSE}
diamonds <- diamonds %>% 
  arrange(price) %>% 
  filter(price > 300) %>% 
  mutate(log_price = log(price))

diamonds
```

--

## Insert with **`ctrl/cmd + shift + m`**

---

## Your turn 6

### Use `%>%` to write a sequence of functions that: 

### 1. Filter only countries that are in the continent of Oceania.
### 2. Select the `country`, `year` and `lifeExp` columns  
### 3. Arrange the results so that the highest life expetency is at the top.

---

## Your turn 6

```{r}
gapminder %>% 
  filter(continent == "Oceania") %>% 
  select(country, year, lifeExp) %>% 
  arrange(desc(lifeExp))
```

---

## The main verbs of `dplyr`

## `select()`

## `filter()`

## `mutate()`

## `arrange()`

## `r font_col("<code>summarize()</code>", "#E69F00")` = `r font_col("Summarize the data", "#56B4E9")`

## `r font_col("<code>group_by()</code>", "#E69F00")` = `r font_col("Group the data", "#56B4E9")`
---

## `summarize()`

```{r, eval=FALSE}
summarize(<DATA>, <NAME> = <FUNCTION>)
```

---

## `summarize()`

```{r}
summarize(diamonds, n = n(), mean_price = mean(price))
```

---

## Your turn 7

## Use `summarise()` to compute three statistics about the gapminder data set:

### 1. The first (`min()`) year in the data
### 2. The last (`max()`) year in the data  
### 3. The total number of observations (`n()`) and the total number of unique countries in the data (`n_distinct()`)

---

## Your turn 7

```{r}
gapminder %>% 
  summarize(
    first = min(year), 
    last = max(year), 
    n = n(), 
    n_countries = n_distinct(country)
  )
```

---

## `group_by()`

```{r, eval=FALSE}
group_by(<DATA>, <VARIABLE>)
```

---

## `group_by()`

```{r, eval = FALSE}
diamonds %>% 
  group_by(cut)
```

---

## `group_by()`

```{r, highlight.output = 2}
diamonds %>% 
  group_by(cut)
```

---

## `group_by()`

```{r, eval = FALSE}
diamonds %>% 
  group_by(cut) %>% #<<
  summarize(n = n(), mean_price = mean(price)) #<<
```

---

## `group_by()`

```{r}
diamonds %>% 
  group_by(cut) %>% 
  summarize(n = n(), mean_price = mean(price))
```

---

## `group_by()`

```{r, eval = FALSE}
diamonds %>% 
  group_by(cut) %>% 
  mutate(n = n(), mean_price = mean(price))
```

---

## `group_by()`

```{r, highlight.output = 1:2}
diamonds %>% 
  group_by(cut) %>% 
  mutate(n = n(), mean_price = mean(price))
```

---

## Your turn 8

## Extract the rows where `continent == "Europe"`. Then use `group_by()` to group by country. Finally, use `summarize()` to compute:

### 1. The total number of observations for each country in Europe
### 2. The lowest observed life expectancy for each country 

---

## Your turn 8

```{r}
gapminder %>% 
  filter(continent == "Europe") %>% 
  group_by(country) %>% 
  summarize(n = n(), min_le = min(lifeExp))
```

---

## Your turn 9

### Use grouping to calculate the mean life expectancy for each continent and year. Call the mean life expectancy variable `mean_le`. Plot the life expectancy over time (no need to change the plot code).

```{r, eval = FALSE}
gapminder %>% 
  __________ %>% 
  __________ %>% 
  ggplot(aes(x = year, y = mean_le, col = continent)) +
    geom_line() + 
    scale_color_manual(values = continent_colors)
```

---

## Your turn 9

### Use grouping to calculate the mean life expectancy for each continent and year. Call the mean life expectancy variable `mean_le`. Plot the life expectancy over time (no need to change the plot code).

```{r, eval = FALSE}
gapminder %>% 
  group_by(continent, year) %>% 
  summarize(mean_le = mean(lifeExp)) %>% 
  ggplot(aes(x = year, y = mean_le, col = continent)) +
    geom_line() + 
    scale_color_manual(values = continent_colors)
```

---

## `mutate_if/all/at()`
## `summarize_if/all/at()`

--

### ... etc!

---

## Joining data

### Use `left_join()`, `right_join()`, `full_join()`, or `inner_join()` to join datasets
### Use `semi_join()` or `anti_join()` to filter datasets against each other

---

class: inverse, center

# Resources
## [R for Data Science](http://r4ds.had.co.nz/): A comprehensive but friendly introduction to the tidyverse. Free online.
## [RStudio Primers](https://rstudio.cloud/learn/primers): Free interactive courses in the Tidyverse
## [10 dplyr tips](https://twitter.com/i/moments/1044926554239881217): a Twitter thread on other useful aspects of dplyr

---

class: inverse, center, middle

### Thank you!

![](https://media.giphy.com/media/XreQmk7ETCak0/giphy.gif)

### `r icon::fa("github")` [malcolmbarrett](https://github.com/malcolmbarrett/)
### `r icon::fa("twitter")` [@malco_barrett](https://twitter.com/malco_barrett)

Slides created via the R package [xaringan](https://github.com/yihui/xaringan).

---

class: inverse, center

# Game plan:
## *Bonus*: `r font_col("joining data and <code>dplyr</code> with databases", "#E69F00")`

---

```{r, include=FALSE}
library(nycflights13)

# Toy datasets to use

superheroes <- tribble(
       ~name, ~alignment,  ~gender,          ~publisher,
   "Magneto",      "bad",   "male",            "Marvel",
     "Storm",     "good", "female",            "Marvel",
  "Mystique",      "bad", "female",            "Marvel",
    "Batman",     "good",   "male",                "DC",
     "Joker",      "bad",   "male",                "DC",
  "Catwoman",      "bad", "female",                "DC",
   "Hellboy",     "good",   "male", "Dark Horse Comics"
  )

publishers <- tribble(
  ~publisher, ~yr_founded,
        "DC",       1934L,
    "Marvel",       1939L,
     "Image",       1992L
  )

```


```{r}
superheroes
```

```{r}
publishers
```

---

## Joining data

```{r}
superheroes %>% 
  left_join(publishers, by = "publisher")
```

---

## Joining data

```{r}
publishers %>% 
  left_join(superheroes, by = "publisher")
```

---

![](https://raw.githubusercontent.com/gadenbuie/tidyexplain/master/images/left-join.gif)

---

![](https://raw.githubusercontent.com/gadenbuie/tidyexplain/master/images/right-join.gif)

---

![](https://raw.githubusercontent.com/gadenbuie/tidyexplain/master/images/full-join.gif)

---

![](https://raw.githubusercontent.com/gadenbuie/tidyexplain/master/images/inner-join.gif)

---

![](https://raw.githubusercontent.com/gadenbuie/tidyexplain/master/images/semi-join.gif)

---

![](https://raw.githubusercontent.com/gadenbuie/tidyexplain/master/images/anti-join.gif)
---

## SQL with dplyr

```{r}
# install.packages("RSQLite")

library(dbplyr)
library(DBI)

con <- DBI::dbConnect(RSQLite::SQLite(), path = ":dbname:")
copy_to(con, gapminder, "gapminder")

gapminder_tbl <- tbl(con, "gapminder")
```

---

```{r}
gapminder_tbl
```

---

## dplyr works out-of-box with numerous types of databases

```{r}
gapminder_tbl %>%
  summarize(
    first = min(year, na.rm = TRUE), 
    last = max(year, na.rm = TRUE), 
    n = n(), 
    n_countries = n_distinct(country)
   )
```

---

## dplyr works out-of-box with numerous types of databases

```{r}
gapminder_tbl %>%
  summarize(
    first = min(year, na.rm = TRUE), 
    last = max(year, na.rm = TRUE), 
    n = n(), 
    n_countries = n_distinct(country)
   ) %>% 
  show_query()
```

--

## <SQL>
### SELECT MIN(`` `year` ``) AS `` `first` ``, MAX(`` `year` ``) AS `` `last` ``, COUNT() AS `` `n` ``, COUNT(DISTINCT `` `country` ``) AS `` `n_countries` ``
### FROM `` `gapminder` `` `

---

## use `collect()` to pull to local data frame

```{r}
local_df <- gapminder_tbl %>%
  summarize(
    first = min(year, na.rm = TRUE), 
    last = max(year, na.rm = TRUE), 
    n = n(), 
    n_countries = n_distinct(country)
   ) %>% 
  collect()

local_df
```

---

# Disconnect

```{r}
dbDisconnect(con)
```

---

class: inverse, center

# Game plan:
## Bonus: joining data and <code>dplyr</code> with databases
## *Bonus*: `r font_col("tidying data: <code>tidyr</code>", "#E69F00")`

```{r lotr, include=FALSE}
lotr <- tribble(
                         ~film,    ~race, ~female, ~male,
  "The Fellowship Of The Ring",    "Elf",   1229L,  971L,
  "The Fellowship Of The Ring", "Hobbit",     14L, 3644L,
  "The Fellowship Of The Ring",    "Man",      0L, 1995L,
              "The Two Towers",    "Elf",    331L,  513L,
              "The Two Towers", "Hobbit",      0L, 2463L,
              "The Two Towers",    "Man",    401L, 3589L,
      "The Return Of The King",    "Elf",    183L,  510L,
      "The Return Of The King", "Hobbit",      2L, 2673L,
      "The Return Of The King",    "Man",    268L, 2459L
  )
```
---

background-image: url(http://hexb.in/hexagons/tidyr.png)
background-position: 90% 10%

## `r font_col("tidyr", "#E69F00")`

## Functions for tidying data. 
## What is tidy data?

> “Tidy datasets are all alike, but every messy dataset is messy in its own way.” –– Hadley Wickham

---

# Tidy Data
![](http://garrettgman.github.io/images/tidy-1.png)</div>

---

# Tidy Data 
![](http://garrettgman.github.io/images/tidy-1.png)</div>
### .medium[Each `r font_col("column", "#E69F00")` is a single `r font_col("variable", "#56B4E9")`]

---

# Tidy Data
![](http://garrettgman.github.io/images/tidy-1.png)</div>
### `r font_col(".medium[Each column is a single variable]", "#E5E5E5")`
### .medium[Each `r font_col("row", "#E69F00")` is a single `r font_col("observation", "#56B4E9")`]

---

# Tidy Data 
![](http://garrettgman.github.io/images/tidy-1.png)</div>
### `r font_col(".medium[Each column is a single variable]", "#E5E5E5")`
### `r font_col(".medium[Each row is a single observation]", "#E5E5E5")`
### .medium[Each `r font_col("cell", "#E69F00")` is a `r font_col("value", "#56B4E9")`]

---

## `gather()`

```{r, eval=FALSE}
gather(<DATA>, "<KEY>", "<VALUE>", <VARIABLES>)
```

---

## `gather()`

```{r}
lotr
```

---

```{r}
lotr %>% 
  gather("sex", "words", female:male)
```

---

```{r}
table4a %>% 
  gather("year", "cases", -country)
```

---

## `spread()`

```{r, eval=FALSE}
spread(<DATA>, <KEY>, <VALUE>)
```

---

## `spread()`

```{r}
lotr %>% 
  gather("sex", "words", female:male) %>% 
  spread(race, words)
```

---

```{r}
table2 %>% 
  spread(type, count) %>% 
  mutate(prevalence = (cases/population) * 100000)
```

---

```{r}
who %>% 
  gather("codes", "n", 5:60) %>% 
  select(country, year, codes, n)
```

---

## `separate()`/`unite()`

```{r, eval=FALSE}
separate(<DATA>, <VARIABLE>, into = c("<VARIABLE1>", "<VARIABLE2>"))

unite(<DATA>, <VARIABLES>)
```

---

```{r}
cases <- tribble(
   ~id,     ~sex_age,
   "1",    "male_56",
   "2",  "female_77",    
   "3",  "female_49"
)
separate(cases, sex_age, into = c("sex", "age"), convert = TRUE)
```
